ERCS/erc-7715.md

```diff
@@ -0,0 +1,551 @@
+---
+eip: 7715
+title: Grant Permissions from Wallets
+description: Adds JSON-RPC method for granting permissions from a wallet
+author: Luka Isailovic (@lukaisailovic), Derek Rein (@arein), Dan Finlay (@danfinlay), Derek Chiang (@derekchiang), Fil Makarov (@filmakarov), Pedro Gomes (@pedrouid), Conner Swenberg (@ilikesymmetry)
+discussions-to: https://ethereum-magicians.org/t/erc-7715-grant-permissions-from-wallets/20100
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-05-24
+requires: 4337, 5792, 7679, 7710
+---
+
+## Abstract
+
+We define a new JSON-RPC method `wallet_grantPermissions` for DApp to request a Wallet to grant permissions in order to execute transactions on the user’s behalf. This enables two use cases:
+
+- Executing transactions for users without a wallet connection.
+- Executing transactions for users with a wallet connection that is scoped with permissions.
+
+## Motivation
+
+Currently most DApps implement a flow similar to the following:
+
+```mermaid
+sequenceDiagram
+    User ->>+ DApp: Establish Connection
+    loop For each call
+    DApp ->> User: Propose Call <br> (Transaction)
+    User ->> DApp: Approve Call
+    end
+```
+
+Each interaction requires the user to sign a transaction with their wallet. The problems are:
+
+- It can get tedious for the user to manually approve every transaction, especially in highly-interactive applications such as games.
+- It’s impossible to send transactions for users without an active wallet connection. This invalidates use cases such as subscriptions, passive investments, limit orders, and more.
+
+In the context of AA, there are multiple vendor-specific implementations of session keys, which are temporary keys imbued with specific permissions. However, since the implementations are vendor-specific, it’s impossible for DApps to “request” session keys from wallets in a unified way, regardless of the specific wallet implementations.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### `wallet_grantPermissions`
+
+We introduce a `wallet_grantPermissions` method for the DApp to request the Wallet to grant permissions.
+
+#### Permission schema
+
+```tsx
+type PermissionRequest = {
+  chainId: Hex; // hex-encoding of uint256
+  address?: Address;
+  expiry: number; // unix timestamp
+  signer: {
+    type: string; // enum defined by ERCs
+    data: Record<string, any>;
+  };
+  permission: {
+    type: string; // enum defined by ERCs
+    data: Record<string, any>;
+  };
+  policies: {
+    type: string; // enum defined by ERCs
+    data: Record<string, any>;
+  }[];
```

Comment:
any reason not to be consistent with `permission` and `policies` either both being an `Array` or both being singletons? 

looking at some examples policies vs permissions below, it's also not super clear what the difference is between them. a `NativeTokenSpendPolicy` seems similar to an `Erc20TokenTransferPermission`. They both define limits on the spend of some funds.

Reply:
When I was first looking at this, I was thinking it would be used like:

```
{
  permission: { type: "owner" },
  policies: [], // no need to for policies because this is a root permission
}
```

or 

```
{
  permission: { type: "session-key", data: { expiry: fifteenMinutesInMs } },
  policies: [
    ... // spend limits, etc
  ]
}
```